File: src/main/scala/akkastudy/supervisorstrategy/scala/ExceptionalChildActor.scala
package akkastudy.supervisorstrategy.scala

import akka.actor.Actor
import akka.event.Logging

class ExceptionalChildActor extends Actor {

  val log = Logging(context.system, this)

  def receive = {
    case "ArithmeticException" => throw new ArithmeticException("Oh no")
    case "NullPointerException" => throw new NullPointerException("Oh no")
    case "IllegalArgumentException" => throw new IllegalArgumentException("Oh no")
    case "Exception" => throw new Exception("Oh no")
    case x @ _ => sender ! s"Message Received: $x"
  }

  @throws[Exception]
  override def postStop {
    log.info("Exceptional Child Actor Stopped")
    super.postStop
  }

  @throws[Exception]
  override def postRestart(reason: Throwable) {
    log.info("Exceptional Child Actor Restarted")
    super.postRestart(reason)
  }
}

File: src/main/java/akkastudy/supervisorstrategy/java/ExceptionalChildActor.java
package akkastudy.supervisorstrategy.java;

import akka.actor.UntypedActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;

public class ExceptionalChildActor extends UntypedActor {

    LoggingAdapter log = Logging.getLogger(getContext().system(), this);

    @Override
    public void onReceive(Object message) throws Exception {
        if (message.equals("ArithmeticException")) throw new ArithmeticException("Oh no");
        else if (message.equals("NullPointerException")) throw new NullPointerException("Oh no");
        else if (message.equals("IllegalArgumentException")) throw new IllegalArgumentException("Oh no");
        else if (message.equals("Exception")) throw new Exception("Oh no");
        else sender().tell(String.format("Message Received %s", message), null);
    }


    @Override
    public void postStop() throws Exception {
        log.info("Exceptional Child Actor Stopped");
        super.postStop();
    }

    public void postRestart(Throwable reason) throws Exception {
        log.info("Exceptional Child Actor Restarted");
        super.postRestart(reason);
    }
}
